### PR TITLE
Implement File and Dir for win32

### PR DESCRIPTION
**Pull request by RX14**
This is fairly WIP, since it depends on #5584 which isn't finalized. It works though.

The main goal is to continue to reduce the amount of stubbing out in `file_spec` and `dir_spec`.

### CODE REVIEW

**Comment by RX14** [comment]
CI failed because of formatting... I'll fix this when #5584 is merged.

File: src/crystal/system/win32/time.cr [hunk 1]
```diff
@@ -26,6 +30,11 @@ module Crystal::System::Time
     {seconds, nanoseconds}
   end
 
+  def self.from_filetime(filetime) : ::Time
+    seconds, nanoseconds = filetime_to_seconds_and_nanoseconds(filetime)
+    ::Time.new(seconds: seconds, nanoseconds: nanoseconds, location: ::Time::Location::UTC)
```
Comment (straight-shoota) [inline src/crystal/system/win32/time.cr]: `::Time.utc`. This time for real ;)

File: spec/std/dir_spec.cr [hunk 2]
```diff
@@ -31,9 +36,13 @@ describe "Dir" do
     end
 
     it "tests empty? on an empty directory" do
-      path = "/tmp/crystal_empty_test_#{Process.pid}/"
-      Dir.mkdir(path, 0o700)
-      Dir.empty?(path).should be_true
+      path = "#{__DIR__}/data/crystal_empty_test/"
+      begin
```
[the remark this answers is not in the document]
Comment (RX14) [reply spec/std/dir_spec.cr]: What needs to happen is we need to provide appropriate tools to cross-compile the spec suite (i.e. never depend on the current directory and be able to tell the spec suite the location of the data and temporary dirs to use) with global helper methods. I think that's neccesary for the whole spec suite which works with files.

File: src/file.cr [hunk 3]
```diff
@@ -769,26 +739,28 @@ class File < IO::FileDescriptor
...
   end
 
-  # Yields an `IO` to read a section inside this file.
-  # Multiple sections can be read concurrently.
-  def read_at(offset, bytesize, &block)
-    self_bytesize = self.size
+  {% unless flag?(:win32) %}
+    # Yields an `IO` to read a section inside this file.
+    # Multiple sections can be read concurrently.
+    def read_at(offset, bytesize, &block)
```
[the remark this answers is not in the document]
Comment (RX14) [reply src/file.cr]: This can work on windows, actually. It needs to be fixed though.

File: spec/std/dir_spec.cr (hunk 2 above)
[the remark this answers is not in the document]
Comment (straight-shoota) [reply spec/std/dir_spec.cr]: Agreed. It's a different thing, though.

**Comment by straight-shoota** [comment]
Now that #5584 got merged, this needs a rebase.

File: src/crystal/system/win32/file_info.cr [hunk 4]
```diff
@@ -0,0 +1,97 @@
...
+      nNumberOfLinks: 1,
+      nFileIndexHigh: 0,
+      nFileIndexLow: 0
+    )
+    @file_type = LibC::FILE_TYPE_DISK
+    @reparse_tag = file_attributes.dwReserved0
+  end
+
+  def initialize(@file_type : LibC::DWORD)
+    # @file_attributes = LibC::BY_HANDLE_FILE_INFORMATION.new(0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0)
```
Comment (RX14) [inline src/crystal/system/win32/file_info.cr]: Remove this.

**Comment by RX14** [comment]
The most useful ideas would be on how to clean up the spec suite to work better on windows. @straight-shoota's tempfile and data directory enhancements would probably be helpful too.

File: spec/std/file_spec.cr [hunk 5]
```diff
@@ -83,27 +84,30 @@ describe "File" do
     idx.should eq(20)
   end
 
-  it "reads lines from file with each as iterator" do
-    idx = 0
-    File.each_line("#{__DIR__}/data/test_file.txt").each do |line|
-      if idx == 0
-        line.should eq("Hello World")
+  {% unless flag?(:win32) %}
+    # TODO: these are broken on win32 because they leak file descriptors which break later tests
```
Comment (RX14) [inline spec/std/file_spec.cr]: This is a huge problem: a file can only be opened once on windows. This test tests the `each_line` iterator, which leaks a filedescriptor and therefore breaks all the tests after it.

I think the best way to solve this is to remove the `File.each_line` iterator (keep the block version) and force people to use

```cr
File.open(...) do |file|
  file.each_line...
end
```

to make leaking the file descriptor harder.

File: spec/std/file_spec.cr [hunk 6]
```diff
@@ -126,7 +130,9 @@ describe "File" do
       ex = expect_raises(Errno, /Error determining size/) do
         File.empty?("#{__FILE__}/")
       end
-      ex.errno.should eq(Errno::ENOTDIR)
+      {% unless flag?(:win32) %}
+        ex.errno.should eq(Errno::ENOTDIR)
```
Comment (RX14) [inline spec/std/file_spec.cr]: These are because `expect_raises` is stubbed out on `win32`, so the return type changes.

File: spec/std/file_spec.cr [hunk 7]
```diff
@@ -329,7 +337,11 @@ describe "File" do
       begin
         File.write(path, "")
         File.chmod(path, 0o775)
-        File.info(path).permissions.should eq(File::Permissions.new(0o775))
+        {% if flag?(:win32) %}
+          File.info(path).permissions.should eq(File::Permissions.new(0o666))
```
Comment (RX14) [inline spec/std/file_spec.cr]: Windows doesn't have the full permissions model, so i'm not sure how to clean up these specs.

File: spec/std/file_spec.cr [hunk 8]
```diff
@@ -405,12 +429,15 @@ describe "File" do
     end
   end
 
-  it "gets info for pipe" do
-    IO.pipe do |r, w|
-      r.info.type.should eq(File::Type::Pipe)
-      w.info.type.should eq(File::Type::Pipe)
+  {% unless flag?(:win32) %}
+    # TODO: Implement IO.pipe for windows
```
Comment (RX14) [inline spec/std/file_spec.cr]: TODO: this is done

File: spec/std/file_spec.cr [hunk 9]
```diff
@@ -507,108 +537,117 @@ describe "File" do
...
-      File.expand_path("a", nil).should eq(File.join([base, "a"]))
-    end
-
-    it "converts a pathname to an absolute pathname, Ruby-Talk:18512" do
-      File.expand_path(".a").should eq(File.join([base, ".a"]))
-      File.expand_path("..a").should eq(File.join([base, "..a"]))
-      File.expand_path("a../b").should eq(File.join([base, "a../b"]))
-    end
+  {% unless flag?(:win32) %}
+    describe "expand_path" do
```
Comment (RX14) [inline spec/std/file_spec.cr]: `expand_path` is utterly broken beyond repair on windows.

File: spec/std/file_spec.cr [hunk 10]
```diff
@@ -507,108 +537,117 @@ describe "File" do
...
     end
-  end
+  {% end %}
 
   describe "real_path" do
     it "expands paths for normal files" do
-      File.real_path("/usr/share").should eq("/usr/share")
-      File.real_path("/usr/share/..").should eq("/usr")
+      {% if flag?(:win32) %}
+        File.real_path("/usr/share").should eq("C:\\usr\\share")
```
Comment (RX14) [inline spec/std/file_spec.cr]: This depends on `Dir.current` being on `C:\\`...

**Comment by RX14** [comment]
Another painful thing is that porting the spec runner to win32 depends on `File` and `Dir`, and obviously the `File` and `Dir` specs in this PR depend on the spec runner. So until we get spec merged, there's specs in the stdlib we can't run.

File: spec/std/file_spec.cr (hunk 5 above)
Comment (Sija) [reply spec/std/file_spec.cr]: Wouldn't be better to raise on windows inside `File.each_line` iterator, leaving this functionality on other OSes intact?

Comment (RX14) [reply spec/std/file_spec.cr]: No, because leaking file descriptors is still a problem on unix, it's just far more of a visible problem on windows.

Comment (straight-shoota) [reply spec/std/file_spec.cr]: Another suitable alternative to `File.each_line(path)` is `File.read_lines(path).each` which reads the lines into memory and iterates over the array.

File: spec/std/file_spec.cr (hunk 9 above)
Comment (straight-shoota) [reply spec/std/file_spec.cr]: would be solved by #5550

Comment (RX14) [reply spec/std/file_spec.cr]: Indeed. That's why I'm pushing for it to be merged.

**Comment by RX14** [comment]
Updated the PR, still WIP because it's failing a few specs, but ready for review.

**Comment by RX14** [comment]
All specs pass now.

File: src/file.cr [hunk 11]
```diff
@@ -2,22 +2,20 @@ require "crystal/system/file"
...
-  {% else %}
-    "/"
-  {% end %}
+  SEPARATOR_STRING = "/"
 
   # :nodoc:
   DEFAULT_CREATE_PERMISSIONS = File::Permissions.new(0o644)
 
+  {% if flag?(:win32) %}
+    DEVNULL = "NUL"
```
Comment (straight-shoota) [inline src/file.cr]: Is this const used anywhere?

Comment (straight-shoota) [reply src/file.cr]: So `:nodoc:`?

Comment (RX14) [reply src/file.cr]: ...no, why would an application not need this?

Comment (straight-shoota) [reply src/file.cr]: Then it should be documented with a proper description.

Comment (RX14) [reply src/file.cr]: Oh yeah, good point. Thanks.

File: spec/std/dir_spec.cr [hunk 12]
```diff
@@ -8,6 +9,10 @@ private def it_raises_on_null_byte(operation, &block)
   end
 end
 
+private def from_win(files)
+  files.map { |f| f.lchop("C:\\") }
+end
```
Comment (ysbaddaden) [inline spec/std/dir_spec.cr]: That should return `files` unless the `:win32` flag is set.

**Comment by straight-shoota** [comment]
@ysbaddaden Try adding `?w=1`to the compare URL, it'll hide lines from the diff with only whitespace changes. Makes reviewing much easier (although you can't directly comment in this view).

**Comment by straight-shoota** [comment]
@ysbaddaden 

* Forward slash `/` is a legal file separator on Windows, it has even supported since DOS times.
* A path starting with `/` or `\` on windows is rooted in the current drive. It's not an absolute path, but rooted.
* This should work using `Dir.tempdir` which is implemented in this PR or even better `with_tempfile` from #5951.

Most of the details for Windows paths are actually covered in #5635, including a large spec suite.

File: spec/std/dir_spec.cr (hunk 2 above)
[the remark this answers is not in the document]
Comment (ysbaddaden) [reply spec/std/dir_spec.cr]: Since we must change this, a helper to create/drop a tmp folder wouldn't hurt, and avoid specs to duplicate a lot of behavior.

```crystal
def with_tmpdir
  tmpdir = ENV["TMPDIR"]? || ENV["TEMP"] || ENV["TMP"] || "/tmp"
  path = File.join(tmpdir, "crystal-stdspec-#{Random.hex}")
  begin
    Dir.mkdir_p(path)
    yield path
  ensure
    FileUtils.rm_rf(path) if Dir.exists?(path)
  end
end

with_tmpdir do |path|
  Dir.mkdir(path, 0o700)
  Dir.empty?(path).should be_true
end
```

That doesn't solve the need for a `$DATADIR` environment variable, but sources and destinations are different matters.

[the remark this answers is not in the document]
Comment (straight-shoota) [reply spec/std/dir_spec.cr]: Thats all already covered in #5951

**Comment by RX14** [comment]
> Also, green specs should mean that everything is implemented (done!). A spec that should eventually pass but doesn't for some reason must fail. Let it fail, or stub the spec body if it can't compile, but make sure to flunk (not skip).

I think i'll make them pending.

And yeah, as @straight-shoota said, windows has implemented fairly "unix-compatible" paths since aparrently MS-DOS 2 (just not at the commandline, since `/` is the option string). The reason i'm not even attempting to try and use true windows paths is because it would make this PR even more horifically large than it is. Before I removed the code that made the seperator `\` on windows, all the path handling commands were entirely broken because they assumed that `\` was *the* seperator on windows. It's not, both `\` and `/` are seperators and so you got `\/\/` monstrosities and all the specs were broken. So until we merge `Path`, and sort this out properly, the specs are essentially a hack. But the hack works pretty well if you only have one drive.

**Comment by ysbaddaden** [comment]
Sigh. Windows is even more horrible and confusing than I could believe.

**Comment by RX14** [comment]
@ysbaddaden just wait until you see how many pages of code it takes to dereference a symlink.

**Comment by RX14** [comment]
Time for another review.

The specs diff is *much* smaller this time, and i'm much happier with this PR (only took 6 months).

File: spec/std/dir_spec.cr [hunk 13]
```diff
@@ -47,7 +47,9 @@ describe "Dir" do
       ex = expect_raises(Errno, /Error determining size of/) do
         Dir.empty?(datapath("dir", "f1.txt", "/"))
       end
-      ex.errno.should eq(Errno::ENOTDIR)
+      {% unless flag?(:win32) %}
```
Comment (RX14) [inline spec/std/dir_spec.cr]: These changes are made because `expect_raises` is stubbed out on windows (in a future PR) and returns `nil`.

File: spec/std/file_spec.cr [hunk 14]
```diff
@@ -27,11 +27,35 @@ private def it_raises_on_null_byte(operation, &block)
...
+    path
+  {% end %}
+end
+
 describe "File" do
   it "gets path" do
     path = datapath("test_file.txt")
-    file = File.new path
-    file.path.should eq(path)
+    File.open(path) do |file|
```
Comment (RX14) [inline spec/std/file_spec.cr]: All these changes from `File.new` to `File.open` are required because files are locked by default on windows, so file descriptor leaks are far more serious (and they're mostly required to make specs pass)

File: spec/std/file_spec.cr [hunk 15]
```diff
@@ -234,11 +262,11 @@ describe "File" do
...
   describe "same?" do
     it "compares following symlinks only if requested" do
-      file = File.expand_path(datapath("test_file.txt"))
-      other = File.expand_path(datapath("test_file.ini"))
+      file = File.real_path(datapath("test_file.txt"))
+      other = File.real_path(datapath("test_file.ini"))
 
       with_tempfile("test_file_symlink.txt") do |symlink|
-        File.symlink(file, symlink)
+        File.symlink(to_windows_path(file), symlink)
```
Comment (RX14) [inline spec/std/file_spec.cr]: Everything handles unix-style paths on windows apart from the *target* of `File.symlink`. Because `Path` PR isn't merged yet, path handling remains broken in windows, so hacks like this are required to get specs to run.

File: src/file.cr [hunk 16]
```diff
@@ -360,9 +358,7 @@ class File < IO::FileDescriptor
     end
 
     String.build do |str|
-      {% if !flag?(:windows) %}
-        str << SEPARATOR_STRING
-      {% end %}
+      str << SEPARATOR_STRING
       items.join SEPARATOR_STRING, str
     end
```
Comment (j8r) [inline src/file.cr]: Why `String.build` vs `SEPARATOR_STRING + items.join SEPARATOR_STRING`?

Comment (asterite) [reply src/file.cr]: `SEPARATOR_STRING + items.join SEPARATOR_STRING` creates two strings (one to join, the other for the `+`). `String.build` will just create one string.

File: spec/std/file_spec.cr [hunk 17]
```diff
@@ -569,53 +608,56 @@ describe "File" do
       File.expand_path(base).should eq(base)
     end
 
-    it "converts a pathname to an absolute pathname, using ~ (home) as base" do
+    pending_win32 "converts a pathname to an absolute pathname, using ~ (home) as base" do
       File.expand_path("~/").should eq(home)
       File.expand_path("~/..badfilename").should eq(File.join(home, "..badfilename"))
-      File.expand_path("..").should eq("/#{base.split('/')[0...-1].join('/')}".gsub(%r{\A//}, "/"))
+      File.expand_path("..").should eq("/#{base.split('/')[0...-1].join('/')}".gsub(/\A\/\//, "/"))
```
Comment (straight-shoota) [inline spec/std/file_spec.cr]: Why these changes?

File: spec/std/file_spec.cr [hunk 18]
```diff
@@ -27,11 +27,35 @@ private def it_raises_on_null_byte(operation, &block)
...
+    normalized_permissions = 0o444
+    normalized_permissions |= 0o222 if permissions.bits_set?(0o200)
+    normalized_permissions |= 0o111 if directory
+    File::Permissions.new(normalized_permissions)
+  {% else %}
+    File::Permissions.new(permissions)
+  {% end %}
+end
+
+private def to_windows_path(path)
```
Comment (straight-shoota) [inline spec/std/file_spec.cr]: Maybe better `to_native_path`?

File: spec/std/spec_helper.cr [hunk 19]
```diff
@@ -4,3 +4,13 @@ require "../support/tempfile"
 def datapath(*components)
   File.join("spec", "std", "data", *components)
 end
+
+{% if flag?(:win32) %}
+  def pending_win32(description = "assert", file = __FILE__, line = __LINE__, end_line = __END_LINE__, &block)
```
Comment (straight-shoota) [inline spec/std/spec_helper.cr]: This is a pretty straightforward implementation, but having all those `pending_win32` in the specs looks a bit strange. The examples are foremost actual specs and should be called `it`. Maybe `it "foobar", win32: :pending` would be a better alternative?

File: spec/support/tempfile.cr [hunk 20]
```diff
@@ -1,7 +1,11 @@
 require "tempfile"
 require "file_utils"
 
-SPEC_TEMPFILE_PATH = File.join(Tempfile.dirname, "cr-spec-#{Random.new.hex(4)}")
+{% if flag?(:win32) %}
+  SPEC_TEMPFILE_PATH = File.join(Tempfile.dirname, "cr-spec-#{Random.new.hex(4)}").gsub("C:\\", '/').gsub('\\', '/')
```
Comment (straight-shoota) [inline spec/support/tempfile.cr]: Why remove drive from path?

File: src/crystal/system/win32/file_info.cr [hunk 21]
```diff
@@ -0,0 +1,96 @@
...
+
+  def same_file?(other : ::File::Info) : Bool
+    return false if type.symlink? || type.pipe? || type.character_device?
+
+    @file_attributes.dwVolumeSerialNumber == other.file_attributes.dwVolumeSerialNumber &&
+      @file_attributes.nFileIndexHigh == other.file_attributes.nFileIndexHigh &&
+      @file_attributes.nFileIndexLow == other.file_attributes.nFileIndexLow
+  end
+
+  private def to_windows_path(path : String) : LibC::LPWSTR
```
Comment (straight-shoota) [inline src/crystal/system/win32/file_info.cr]: It's trivial, but could maybe reuse `Crystal::System::Dir.to_windows_path`, optionally moved to a different location.

File: spec/support/tempfile.cr (hunk 20 above)
Comment (RX14) [reply spec/support/tempfile.cr]: Because `FileUtils.mkdir_p` can't handle windows paths.

File: src/crystal/system/win32/file_info.cr (hunk 21 above)
Comment (RX14) [reply src/crystal/system/win32/file_info.cr]: A little bit of copying never did no harm, at least for a tiny method like this.

File: spec/std/spec_helper.cr (hunk 19 above)
Comment (RX14) [reply spec/std/spec_helper.cr]: That's overthinking it. Unless you want to make a pull request to `Spec` itself I see no reason to make this look super fancy when all these pendings will be removed sooner or later.

File: spec/std/file_spec.cr (hunk 17 above)
Comment (RX14) [reply spec/std/file_spec.cr]: These used to be inside a `{% if flag...`, and then I found out about https://github.com/crystal-lang/crystal/issues/6180.

Not neccesary any more because they're not in a macro.

File: spec/std/dir_spec.cr [hunk 22]
```diff
@@ -230,15 +232,18 @@ describe "Dir" do
...
 
     it "tests with relative path (starts with ..)" do
-      base_path = File.join("..", File.basename(Dir.current), "spec", "std", "data", "dir")
-      Dir["../#{File.basename(Dir.current)}/#{datapath}/dir/*/"].sort.should eq [
-        File.join(base_path, "dots", ""),
-        File.join(base_path, "subdir", ""),
-        File.join(base_path, "subdir2", ""),
-      ].sort
+      Dir.cd(datapath) do
+        base_path = "../../../spec/std/data/dir"
```
[the remark this answers is not in the document]
Comment (RX14) [reply spec/std/dir_spec.cr]: Why not? Thats closest to how it originally was.

[the remark this answers is not in the document]
Comment (straight-shoota) [reply spec/std/dir_spec.cr]: This example is about having a path start with `..`. It doesn't need multiple. Besides, this path makes assumptions about the location of `datapath` which is not strictly wrong, but can be avoided.

My suggestion:
```cr
Dir.cd(datapath("dir")) do
  base_path = "../dir"
  # ...
```

This is much shorter and clearly focuses the purpose of this spec.

File: spec/std/spec_helper.cr (hunk 19 above)
Comment (straight-shoota) [reply spec/std/spec_helper.cr]: Yeah, that's probably not worth it changing `it`. But maybe there is a better name for this method...?

File: spec/support/tempfile.cr (hunk 20 above)
Comment (straight-shoota) [reply spec/support/tempfile.cr]: How about adding a TODO?

Comment (straight-shoota) [reply spec/support/tempfile.cr]: It would ne nice to have something like `.gsub(/\A[a-zA-Z]:\\\\/)` which works with a temp path on other drives than `C:`.

File: spec/std/dir_spec.cr (hunk 22 above)
[the remark this answers is not in the document]
Comment (RX14) [reply spec/std/dir_spec.cr]: Yeah, if you're fine with it, IIRC you wrote this spec in the first place.

I saw the original spec went through a few hoops to get more than one `../` in the path, so I did too.

File: spec/std/spec_helper.cr (hunk 19 above)
Comment (RX14) [reply spec/std/spec_helper.cr]: Nah, it's clear enough and easy enough to look up. We're hardly writing public API here, code quality and naming of methods in spec files is really not worth bikeshedding about since it's so easy to change.

File: spec/support/tempfile.cr (hunk 20 above)
Comment (RX14) [reply spec/support/tempfile.cr]: The TODO is to merge #5550. And honestly, I'm just hacking the specs around until they work here, then cleaning it up enough to be good to merge. At the end of it all we can just `grep -R 'flag\?(:win32' src` and find all this junk once we merge proper path handling support.

File: spec/std/file_spec.cr (hunk 18 above)
Comment (RX14) [reply spec/std/file_spec.cr]: This is also temporary, should be fixed up in whatever PR we add proper windows path support, hopefully #5550. So I don't think it matters.

File: spec/std/dir_spec.cr (hunk 22 above)
[the remark this answers is not in the document]
Comment (straight-shoota) [reply spec/std/dir_spec.cr]: Yes please. The original spec was just `Dir["../crystal/spec/std/data/dir/*"]`. Changing the current dir to `datapath("dir")` simplifies this a lot.

**Comment by RX14** [comment]
Ready for another review

**Comment by RX14** [comment]
Ready for another review

File: src/file.cr [hunk 23]
```diff
@@ -2,22 +2,31 @@ require "crystal/system/file"
 
 class File < IO::FileDescriptor
   # The file/directory separator character. `'/'` in Unix, `'\\'` in Windows.
```
Comment (r00ster91) [inline src/file.cr]: Shouldn't this comment be changed since `/` is a legal file separator in Unix and Windows?

Comment (RX14) [reply src/file.cr]: Yeah, these should be changed to just say they're always `/`. This will be changed anyway once we have `Path`.

I'll wait until more comments arrive to fix this.

File: src/file.cr [hunk 24]
```diff
@@ -10,6 +10,17 @@ class File < IO::FileDescriptor
   # :nodoc:
   DEFAULT_CREATE_PERMISSIONS = File::Permissions.new(0o644)
 
+  # The name of the null device on the host platform. /dev/null on UNIX and NUL
```
Comment (straight-shoota) [inline src/file.cr]: Wrap the names in backticks?

File: src/crystal/system/win32/file.cr [hunk 25]
```diff
@@ -0,0 +1,241 @@
...
+    win_path = to_windows_path(path)
+
+    System.retry_wstr_buffer do |buffer, small_buf|
+      len = LibC.GetFullPathNameW(win_path, buffer.size, buffer, nil)
+      if 0 < len < buffer.size
+        return String.from_utf16(buffer[0, len])
+      elsif small_buf && len > 0
+        next len
+      else
+        raise WinError.new("Error resolving real path of #{path.inspect}")
```
[the remark this answers is not in the document]
Comment (straight-shoota) [reply src/crystal/system/win32/file.cr]: Yes, a new approach following #5003 would be great!

File: spec/std/file_spec.cr [hunk 26]
```diff
@@ -27,11 +27,23 @@ private def it_raises_on_null_byte(operation, &block)
   end
 end
 
+private def normalize_permissions(permissions, *, directory)
```
[the remark this answers is not in the document]
Comment (RX14) [reply spec/std/file_spec.cr]: Windows doesn't have the concept of "execute permissions" (at least not at this level of the API)

File: spec/std/file_spec.cr [hunk 27]
```diff
@@ -82,7 +94,8 @@ describe "File" do
     idx.should eq(20)
   end
 
-  it "reads lines from file with each as iterator" do
+  # TODO: following two specs are broken on win32 because they leak file descriptors which break later tests
```
[the remark this answers is not in the document]
Comment (RX14) [reply spec/std/file_spec.cr]: No, ensure works fine you just can't assume that the GC runs between these specs and others. Relying on `File#ensure` is a hack. Regardless, the purpose of #6301 was to remove these methods because they always leak file references. And that's merged so this is all moot since these specs are gone.

**Comment by RX14** [comment]
> Supporting exceptions and ENV should come before this PR IMO

I think that's perfectionism. This works now and the spec diff is minimal. Bikeshedding about the order is moot when the end result will be the same.

**Comment by bararchy** [comment]
I'm with @RX14 especially now that @txe dropped working on exceptions, we should merge W/E is ready so we won't discourage people from working on features, the more waiting we do the more burnout people become

**Comment by ysbaddaden** [comment]
@bcardiff I thought the same, exceptions should come first, but I think merging this pull request is okay. It's isolated, as it implements target specific stuff, it doesn't break anything and is one step forward supporting windows.

Let's merge, once ready.

File: spec/std/file_spec.cr (hunk 26 above)
[the remark this answers is not in the document]
Comment (bcardiff) [reply spec/std/file_spec.cr]: Yeah I don't expect to do the full ACL.

But either the feature is not enabled/tested on windows, or
there is some translation in the implementation to map (at some extent) the result from windows to the stdlib permission value and have a unified API in both platform.

Neither of these options requires a `normalize_permission` in the specs.

[the remark this answers is not in the document]
Comment (RX14) [reply spec/std/file_spec.cr]: It maps incoming permissions just fine, the problem is that it's a lossy transform, so that when the spec suite comes back and queries the permissions it just set on file xyz it gets different permissions to what it just set back.

**Comment by RX14** [comment]
Final few reviewes fixed up and rebased? FInal review for merge?

**Comment by bcardiff** [comment]
Aren't we waiting on #5635 before merging this PR?

**Comment by RX14** [comment]
No, its orthogonal.

**Comment by bararchy** [comment]
Then LGTM

**Comment by sdogruyol** [comment]
Finally :tada:
